accounts-db/src/accounts_index.rs

```diff
@@ -1110,14 +1110,14 @@ impl<T: IndexValue, U: DiskIndexValue + From<T> + Into<T>> AccountsIndex<T, U> {
         F: FnMut(&Pubkey, (&T, Slot)),
```

Comment:
we already expect all calls to `do_scan_secondary_index` to be slow. This should only be used by rpc. this should be faster than the previous impl, but it will hold the index read lock much longer.

Reply:
> I think I'm missing something. If this impl should be faster, why will it hold the index read lock much longer?

we used to do:
acquire disk index bin read lock, clone read entry, drop disk index bin lock
get read lock on slot list of arc entry
...call `func`

Now, we:
acquire disk index bin read lock
...get read lock on slot list of arc entry
... ... call `func`
...drop read lock on slot list
drop disk index bin read lock

removing the clone and the allocs associated with the self ref struct will be faster
This will make the happy case faster
But, this means that the disk index bin read lock may be held longer than it was held before. Depends on the speed of `func` relative to cloning the arc and cost of creating and dropping the self ref struct we're eliminating.
since `func` is passed to us, we can't know here which is more expensive.

Reply:
That pattern gives us the same locking behavior as the self ref struct. And probably slightly better performance based on @Lichtso 's perf results based on not allocating the self ref struct.
It means all entries we find will be added to the in-mem index (`get_and_then` can avoid adding).
It means we will hold the bin read lock less time and still clone the arc (and drop the clone).
I think it depends on the usage. If the caller is a quick in and out, then probably better to use the new locking behavior. If the caller is expected to be longer, than maybe the clone version you wrote? I think these can all be the new behavior (longer possible index bin lock time).

Reply:
I mainly wanted to make sure we all understood the nuance of locking behavior changes in our minds so we can think more deeply and predict side effects or know to look for them.

Reply:
> Do you think it would be better to do:
> 
> lock index bin clone arc of the index entry unlock index bin lock slot list call `func` unlock slot list

isn't this the impl of `pub fn get_cloned(&self, pubkey: &Pubkey) -> Option<AccountMapEntry<T>> {`